### PR TITLE
[Concept Entry] SwiftUI Tables

### PR DESCRIPTION
**Pull request by arirays**
### Description

- Added SwiftUI Tables concept entry
- Closes issue #2105

### Type of Change

- Adding a new entry
- Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

File: content/swiftui/concepts/tables/tables.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+Tags:
+  - 'iOS'
+  - 'SwiftUI'
+  - 'Tables'
+CatalogContent:
+  - 'learn-ios-course'
+  - 'paths/ios-developer'
+---
+
+
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:16]:
```suggestion
```
No need to give 2 white space here

File: content/swiftui/concepts/tables/tables.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+**Table** in SwiftUI is a structure that allows us to create and display collections of data in columns and rows. Many datasets have a tabular structure and can be efficiently presented in a table format. If the number of rows exceeds the available space, `Table` will enable vertical scrolling automatically. Additionally, on macOS, horizontal scrolling is available when the number of columns exceeds the view's width.
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:17]:
```suggestion
**`Table`** in SwiftUI is a container that enables the creation and presentation of data collections in a tabular format, organized into columns and rows. It is a valuable tool for displaying datasets that have a structured, grid-like layout. When the number of rows exceeds the available vertical space, the `Table` component will automatically enable vertical scrolling to ensure all data is accessible. Moreover, on macOS, horizontal scrolling becomes available when the number of columns exceeds the width of the view.
```

File: content/swiftui/concepts/tables/tables.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+---
+
+
+**Table** in SwiftUI is a structure that allows us to create and display collections of data in columns and rows. Many datasets have a tabular structure and can be efficiently presented in a table format. If the number of rows exceeds the available space, `Table` will enable vertical scrolling automatically. Additionally, on macOS, horizontal scrolling is available when the number of columns exceeds the view's width. 
+
+## Creating a Table 
+
+`Table` is a great option for displaying a collection of data in a compact form. In the example below, in order to create a simple three-column table from an array of Singer your data model has to conform to `Identifiable`.
+
+### Example
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:23]:
```suggestion
## Example

`Table` is an excellent choice for presenting data in a concise, organized manner. In the example below, to create a straightforward three-column table from an array of `Singer`, the data model needs to conform to `Identifiable`.

```
Let's remove the above heading and add it below the example heading only. Also refrain from using pronouns.

File: content/swiftui/concepts/tables/tables.md [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    let ranking: Int
+    let id = UUID()
+    
+    var fullName: String { name + " " + lastName}
+}
+```
+
+## Adding sorting to a table
+
+To make the columns of a `Table` sortable we need to make an array of KeyPathComparator objects with the default sorting inside.  Additionally, provide a binding to an array of sort comparators to watch for changes in the sort order. The table doesn't sort data by itself. To enable sorting, you can monitor changes by using the `.onChange(of:perform:)` modifier and then sort the data within the closure.
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:57]:
```suggestion
To enable sorting of `Table` columns, an array of `KeyPathComparator` objects should be created with default sorting inside. Additionally, a binding to an array of sort comparators needs to be provided to track changes in the sorting order. The table itself does not perform data sorting itself. To activate sorting, monitoring changes can be achieved using the `.onChange(of:perform:)` modifier and then sorting the data within the closure.
```

File: content/swiftui/concepts/tables/tables.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    let id = UUID()
+    
+    var fullName: String { name + " " + lastName}
+}
+```
+
+## Adding sorting to a table
+
+To make the columns of a `Table` sortable we need to make an array of KeyPathComparator objects with the default sorting inside.  Additionally, provide a binding to an array of sort comparators to watch for changes in the sort order. The table doesn't sort data by itself. To enable sorting, you can monitor changes by using the `.onChange(of:perform:)` modifier and then sort the data within the closure. 
+Here's an example of how to add sorting to the previous code:
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:58]:
```suggestion

Here's an example of how to add sorting to the previous code:
```

File: content/swiftui/concepts/tables/tables.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            people.sort(using: newOrder)
+        }
+    }
+}
+```
+
+## Adding selection to a table
+
+ To enhance the interactivity of tables we can enable users to choose one or more rows. To achieve this pass the `Table` a binding to an optional identifier of the item. To create a table that allows for multiple selections, you can establish a binding to a Set of IDs. 
+ Here's an example of how to use multi-select functionality into the previous example:
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:90]:
```suggestion
To improve table interactivity, enabling users to select one or more rows can be achieved. To do this, binding to an optional identifier of the item should be provided to the `Table`. For creating a table that permits multiple selections, a binding to a Set of IDs can be established. Here's an example of how to incorporate multi-select functionality into the previous example:
```

File: content/swiftui/concepts/tables/tables.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+```swift
+ // @State private var selection: Singer.ID? allows one row selection
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:93]:
```suggestion
```

File: content/swiftui/concepts/tables/tables.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+ @State private var selection = Set<Singer.ID>()
+
+ var body: some View {
+    Table(people, selection: $selection, sortOrder: $sortOrder) {...}
+ }
+```
+## Customizing a Table style
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:100]:
```suggestion

## Customizing a Table style
```
Add a line before starting a new heading

File: content/swiftui/concepts/tables/tables.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+One of the easiest methods to change a table's appearance is by using the `.tableStyle(_:)` modifier. This modifier requires one argument, a TableStyle, which defines the style to use for the table. For instance, SwiftUI provides several table styles, such as `InsetTableStyle` and, on macOS, `BorderedTableStyle`. The default style is `AutomaticTableStyle`, which is available on all platforms that support `Table`.
+
+```swift
+Table(people, selection: $selection, sortOrder: $sortOrder) {
+    // Table columns
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:106]:
```suggestion
```

File: content/swiftui/concepts/tables/tables.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+} 
+.tableStyle(InsertTableStyle())
+```
+
+## Setting a fixed width for the table column
+
+You can set a specific width for a table column:
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:113]:
```suggestion
Defining a precise width for a table column can be accomplished as follows:
```

**Review by dakshdeepHERE — CHANGES_REQUESTED** [review]
Here are some initial changes commit these so we can move forward in the review process. 🚀

**Comment by arirays** [comment]
Hi @SSwiniarski I created a PR for issue #3800. Could you please review it? Thank you.

File: content/swiftui/concepts/tables/tables.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+            TableColumn("Ranking") {
+                people in
+                Text(String(people.ranking))
+            }
+        }
+        .onChange(of: sortOrder) { newOrder in
+            people.sort(using: newOrder)
+        }
+    }
+}
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:83]:
```suggestion
struct ContentView: View {
  @State private var people = [
      Singer(name: "Taylor", lastName: "Swift", ranking: 95),
      Singer(name: "Olivia", lastName: "Rodrigo", ranking: 80),
      Singer(name: "Billie", lastName: "Eilish", ranking: 70),
      Singer(name: "Katy", lastName: "Perry", ranking: 75)
  ]
  
  @State private var sortOrder = [KeyPathComparator(\Singer.name)]
  
  var body: some View {
      Table(people, sortOrder: $sortOrder) {
          TableColumn("Name", value: \.name)
          TableColumn("Last Name", value: \.lastName)
          TableColumn("Ranking") {
              people in
              Text(String(people.ranking))
          }
      }
      .onChange(of: sortOrder) { newOrder in
          people.sort(using: newOrder)
      }
  }
}
```

File: content/swiftui/concepts/tables/tables.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+### Adding selection to a table
+
+To improve table interactivity, enabling users to select one or more rows can be achieved. To do this, binding to an optional identifier of the item should be provided to the `Table`. For creating a table that permits multiple selections, a binding to a Set of IDs can be established. Here's an example of how to incorporate multi-select functionality into the previous example:
+
+```swift
+ @State private var selection = Set<Singer.ID>()
+
+ var body: some View {
+    Table(people, selection: $selection, sortOrder: $sortOrder) {...}
+ }
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/tables/tables.md:95]:
```suggestion
 @State private var selection = Set<Singer.ID>()

var body: some View {
  Table(people, selection: $selection, sortOrder: $sortOrder) {...}
}
```

**Review by dakshdeepHERE — CHANGES_REQUESTED** [review]
Few more changes.

**Comment by dakshdeepHERE** [comment]
Hello @arirays, can commit the changes so we can move forward to 2nd review?

**Comment by dakshdeepHERE** [comment]
@arirays Looks good for 2nd review. 🚀

File: content/swiftui/concepts/tables/tables.md [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+Tags:
+  - 'iOS'
+  - 'SwiftUI'
+  - 'Tables'
+CatalogContent:
+  - 'learn-ios-course'
+  - 'paths/ios-developer'
+---
+
+**`Table`** in SwiftUI is a container that enables the creation and presentation of data collections in a tabular format, organized into columns and rows. It is a valuable tool for displaying datasets that have a structured, grid-like layout. When the number of rows exceeds the available vertical space, the `Table` component will automatically enable vertical scrolling to ensure all data is accessible. Moreover, on macOS, horizontal scrolling becomes available when the number of columns exceeds the width of the view.
```
Comment (caupolicandiaz) [inline content/swiftui/concepts/tables/tables.md:16]:
```suggestion
A **`Table`** in SwiftUI is a container that enables the creation and presentation of data collections in a tabular format: values organized into columns and rows. It is a valuable tool for displaying datasets that have a structured, grid-like layout. When the number of rows exceeds the available vertical space, the `Table` component will automatically enable vertical scrolling to ensure all data is accessible. Moreover, on macOS, horizontal scrolling becomes available when the number of columns exceeds the width of the view.
```

File: content/swiftui/concepts/tables/tables.md [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
```
Comment (caupolicandiaz) [inline content/swiftui/concepts/tables/tables.md:17]: Before any examples there should be a syntax section to demonstrate how the structure is implemented generically with pseudo code. For example - 
https://www.codecademy.com/resources/docs/swiftui/views/button

File: content/swiftui/concepts/tables/tables.md [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    let name: String
+    let lastName: String
+    let ranking: Int
+    let id = UUID()
+    
+    var fullName: String { name + " " + lastName}
+}
+```
+
+### Adding sorting to a table
```
Comment (caupolicandiaz) [inline content/swiftui/concepts/tables/tables.md:53]:
```suggestion
### Adding Sorting to a Table
```

File: content/swiftui/concepts/tables/tables.md [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+          }
+      }
+      .onChange(of: sortOrder) { newOrder in
+          people.sort(using: newOrder)
+      }
+  }
+}
+```
+
+### Adding selection to a table
```
Comment (caupolicandiaz) [inline content/swiftui/concepts/tables/tables.md:86]:
```suggestion
### Adding Selection to a Table
```

File: content/swiftui/concepts/tables/tables.md [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+```swift
+ @State private var selection = Set<Singer.ID>()
+
+var body: some View {
+  Table(people, selection: $selection, sortOrder: $sortOrder) {...}
+}
+```
+
+### Customizing a Table style
```
Comment (caupolicandiaz) [inline content/swiftui/concepts/tables/tables.md:98]:
```suggestion
### Customizing a Table Style
```

File: content/swiftui/concepts/tables/tables.md [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+
+One of the easiest methods to change a table's appearance is by using the `.tableStyle(_:)` modifier. This modifier requires one argument, a TableStyle, which defines the style to use for the table. For instance, SwiftUI provides several table styles, such as `InsetTableStyle` and, on macOS, `BorderedTableStyle`. The default style is `AutomaticTableStyle`, which is available on all platforms that support `Table`.
+
+```swift
+Table(people, selection: $selection, sortOrder: $sortOrder) {
+} 
+.tableStyle(InsertTableStyle())
+```
+
+### Setting a fixed width for the table column
```
Comment (caupolicandiaz) [inline content/swiftui/concepts/tables/tables.md:108]:
```suggestion
### Setting a Fixed Width for a Table Column
```

File: content/swiftui/concepts/tables/tables.md [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+ @State private var selection = Set<Singer.ID>()
+
+var body: some View {
+  Table(people, selection: $selection, sortOrder: $sortOrder) {...}
+}
+```
+
+### Customizing a Table style
+
+One of the easiest methods to change a table's appearance is by using the `.tableStyle(_:)` modifier. This modifier requires one argument, a TableStyle, which defines the style to use for the table. For instance, SwiftUI provides several table styles, such as `InsetTableStyle` and, on macOS, `BorderedTableStyle`. The default style is `AutomaticTableStyle`, which is available on all platforms that support `Table`.
```
Comment (caupolicandiaz) [inline content/swiftui/concepts/tables/tables.md:100]:
```suggestion
One of the easiest methods to change a table's appearance is by using the `.tableStyle(_:)` modifier. This modifier requires just one argument, a predefined table style. SwiftUI provides several options, such as `InsetTableStyle` and, on macOS, `BorderedTableStyle`. The default style is `AutomaticTableStyle`, which is available on all platforms that support `Table`.
```

**Review by caupolicandiaz — CHANGES_REQUESTED** [review]
@arirays I've added my initial comments. Please update the entry, and tag me when you've finished making changes. Thanks

File: documentation/concept-entry-template.md
Comment (SSwiniarski) [inline documentation/concept-entry-template.md:1]: @arirays In order to merge this you need to undo the changes to this file. Easiest would be to replace it with a copy of the `content-entry-template.md` from your main branch.

**Review by SSwiniarski — CHANGES_REQUESTED** [review]
@arirays There are edits you made to the documentation you need to revert in order for this to merge.

File: documentation/concept-entry-template.md
Comment (arirays) [reply documentation/concept-entry-template.md:1]: k, @SSwiniarski can I use git reset [commit ID]?

Comment (arirays) [reply documentation/concept-entry-template.md:1]: I did git revert [commit ID]

Comment (SSwiniarski) [reply documentation/concept-entry-template.md:1]: @arirays this is still a problem. Did you try my suggested fix?

Comment (arirays) [reply documentation/concept-entry-template.md:1]: Yes I reverted my concept-entry-template.md by copying it from the main branch.

**Review by caupolicandiaz — APPROVED** [review]
@arirays I made a few direct edits. The filepath for the image was incorrect. All images are stored in an existing folder named media. I updated the path and relocated the file, and updated some of the formatting in the entry. This is ready to merge

**Review by dakshdeepHERE — APPROVED** [review]

**Review by SSwiniarski — APPROVED** [review]

**Comment by arirays** [comment]
Hi @SSwiniarski and @caupolicandiaz after checking out my Codecademy profile I noticed that my contributions are anonymous....Is it possible to add my profile afterwards? I'd appreciate your help or advice, thank you so much!

**Comment by dakshdeepHERE** [comment]
@arirays have you linked your Github account to your Codecademy account? Also, make your Codecademy account public. Don't worry it will update automatically.

**Comment by arirays** [comment]
yes I linked my account to Codecademy account and it is public. Thank you for the reply, hopefully it'll update soon 👍

**Pull request merged**